File: PIGBANK.cpp
#include<iostream>
#include<string.h>

using namespace std;

struct coin
{
	int weight;
	int value;
};


int min(int x, int y)
{
	if(y<0)
	return x;
	else if(y<=x)
	return y;
	else 
	return x;
}


int n;
int dp[10005];
int getMin(int c,struct coin coins[])
{
	for(int i=1;i<=c;i++)
	{
		dp[i]=0x3f3f3f3f;
		for(int j=0;j<n;j++)
		{
			if(i-coins[j].weight>=0)
			dp[i]=min(dp[i],coins[j].value+dp[i-coins[j].weight]);
		}
	}
	return dp[c];
}
int main()
{
	int t,e,f,c,ans;
	scanf("%d",&t);
	while(t--)
	{
		scanf("%d %d",&e,&f);
		c=f-e;
		scanf("%d",&n);
		struct coin coins[n];
		dp[0]=0;
		for(int i=0;i<n;i++)
		{ 
			scanf("%d %d",&coins[i].value,&coins[i].weight);
		}
		ans=getMin(c,coins);
		if(ans==0x3f3f3f3f)
		printf("This is impossible.\n");
		else
		printf("The minimum amount of money in the piggy-bank is %d.\n",ans);
	}
}

File: PIGBANK.java
import java.io.OutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.InputStream;

/**
 * Built using CHelper plug-in
 * Actual solution is at the top
 */
public class Main {
    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        PrintWriter out = new PrintWriter(outputStream);
        PIGBANK solver = new PIGBANK();
        solver.solve(1, in, out);
        out.close();
    }

    static class PIGBANK {
        public void solve(int testNumber, InputReader in, PrintWriter out) {
            int t = in.nextInt();
            for (int i = 0; i < t; i++) {
                int e = in.nextInt();
                int f = in.nextInt();
                int w = f - e;
                int n = in.nextInt();
                int[] value = new int[n];
                int[] weight = new int[n];
                for (int j = 0; j < n; j++) {
                    value[j] = in.nextInt();
                    weight[j] = in.nextInt();
                }
                long[] ans = new long[w + 1];
                Arrays.fill(ans, 500000044);
                ans[0] = 0;
                for (int j = 1; j <= w; j++) {
                    for (int k = 0; k < n; k++) {
                        if (weight[k] <= j) {
                            ans[j] = Math.min(ans[j], value[k] + ans[j - weight[k]]);
                        }
                    }
                }
                if (ans[w] != 500000044) {
                    out.println("The minimum amount of money in the piggy-bank is " + ans[w] + ".");
                } else {
                    out.println("This is impossible.");
                }
            }
        }

    }

    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;

        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (Exception e) {
                    return "";
                }
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }

    }
}

